src/psyclone/psyir/nodes/operation.py

```diff
@@ -224,38 +230,137 @@ def create(operator, lhs, rhs):
             are not of the expected type.
 
         '''
-        if not isinstance(operator, Enum) or \
-           operator not in BinaryOperation.Operator:
+        if (not isinstance(operator, Enum) or
+                operator not in BinaryOperation.Operator):
             raise GenerationError(
                 f"operator argument in create method of BinaryOperation class "
                 f"should be a PSyIR BinaryOperation Operator but found "
                 f"'{type(operator).__name__}'.")
-
         binary_op = BinaryOperation(operator)
         binary_op.addchild(lhs)
         binary_op.addchild(rhs)
         return binary_op
 
+    @property
+    def datatype(self):
+        '''
+        :returns: the datatype of the result of this BinaryOperation.
+        :rtype: :py:class:`psyclone.psyir.symbols.DataType`
+
+        :raises InternalError: if the types of the operands are not permitted
+                               for the given operator.
+        :raises NotImplementedError: if the operands are of the same intrinsic
+                                     type but different precisions.
+        :raises TypeError: if the operands are both arrays but are of different
+                           shapes.
+        '''
+        # Get the types of the operands.
+        argtypes = []
+        for child in self.children:
+            # If the operand is itself an operation this will recurse.
+            dtype = child.datatype
+            if isinstance(dtype, UnknownFortranType):
+                if dtype.partial_datatype:
+                    # We are still OK provided we have partial type information
+                    # since that means the intrinsic type can be handled in the
+                    # PSyIR.
+                    dtype = dtype.partial_datatype
+                else:
+                    return UnknownFortranType("")
+            if isinstance(dtype, DeferredType):
+                # If either operand is of DeferredType then we can't do
+                # any better.
+                return DeferredType()
+            if isinstance(dtype, ArrayType):
+                # We know this is an array but do we know its intrinsic type?
+                if isinstance(dtype.intrinsic, DeferredType):
+                    return DeferredType()
+                if isinstance(dtype.intrinsic, UnknownType):
+                    if (isinstance(dtype.intrinsic, UnknownFortranType) and
+                            dtype.intrinsic.partial_datatype):
+                        dtype = ArrayType(dtype.intrinsic.partial_datatype,
+                                          shape=dtype.shape)
+                    else:
+                        return UnknownFortranType("")
+
+            argtypes.append(dtype)
+
+        base_type = None
+
+        if self.operator not in self._numeric_ops:
+            # Must be a relational or logical operator. Intrinsic type of
+            # result will be boolean.
+            base_type = BOOLEAN_TYPE
```

Comment:
This could be at the top with `return BOOLEAN_TYPE` so we don't need to recurse to the children datatypes nor deal with their complications.

Reply:
Yes, I had it that way originally but then realised that one can have an operation that results in an array of booleans.